thriftrw/spec/common.pyx

```diff
@@ -34,6 +34,24 @@ def fields_eq(fields):
     return __eq__
 
 
+def fields_hash(fields):
+    """Generates an ``__hash__`` method.
+
+    :param list fields:
+        List of fields of the object to be hashed.
+    """
+    def hashable(obj):
+        try:
+            hash(obj)
+        except TypeError:
```

Comment:
Instead of checking hashability of fields at runtime, we can probably figure
that out in advance.

There's a fixed number of type specs, and each of them knows whether it's
hashable: all primitive types and enums are hashable, lists, maps and sets
aren't, and structs and unions are hashable only if all their elements are
hashable.